docs/hacking/coding_best_practices.md

```diff
@@ -0,0 +1,280 @@
+confusing or relies on problematic techniques. When submitting a contribution
+to any open-source project it is worth keeping in mind that it has a potential
+of living in the codebase for a long, sometimes very long time, and during this
+time it might become necessary to examine or slightly modify it more than once.
+It is very worth it to invest some effort into helping future maintainers by
+writing clear and well documented (preferably self-documenting) code. This
+small guide summarizes some best known practices and conventions which cannot
+be caught with a linter.
+
+> As an illustration to the above statement consider the following bit of code:
+> ```
+> &>:1-:v v *_$.@
+>  ^    _$>\:^
+> ```
+> It is pretty hard to tell not just whether the code is correct, but even what is
+> supposed to do (even when one knows the language). On the other hand
+> ```
+> factorial :: Integer -> Integer
+> factorial n = product [1..n]
+> ```
+> is quite easy to comprehend without knowing much details of the underlying language.
+> The ultimate goal is to write boring code that is as clear and straightforward as the
+> second example.
+
+The main goal of a reviewer is to make maintainers and future contributors
+lives easier. Code reviews are done and requests for modifications are issued
+not to flex or to cheaply show off superiority, but to ensure that the code
+meets standards: is clear enough and well documented to be maintained, because
+it is usually up to maintainers to deal with any bugs and inconsistencies in
+contributed components.  All reviews are done in good faith and are
+well-meaning, but prioritise long-term project stability.
+
+Things that work in a individually maintained codebase do not work as well in a
+collaborative project: code has to be clear enough for any engineer more or
+less familiar with the codebase in general to modify it as fast as possible.
+This requires a rather high level of input quality control, otherwise the
+codebase would quickly devolve into a bug generator which is extremely
+expensive to maintain in any semblance of working order.
+
+Code author is usually the most knowledgeable subject matter expert when it
+comes to specifics of workings and intent of their contribution. Procedures and
+transformations which are trivial and obvious to them are very likely not
+trivial and not obvious to everyone else. To ensure that the contributed code
+remains used and useful it is important to clearly and succinctly share this
+knowledge. Thus maintainers will likely ask to share all missing details or
+even to restructure contributed code in such a way that it becomes clear what
+it is doing.
+
+All stated above is especially true for security-focused projects where bugs
+and omissions could become very costly. The tolerances of a security project
+are usually much tighter than of most others since the price of failure is much
+greater.
+
+## Best practices
+
+### Document your decisions
```

Comment:
Should we drop the [comment guidelines](https://github.com/containerbuildsystem/cachi2/blob/main/CONTRIBUTING.md#comment-guidelines) from CONTRIBUTING.md?

Reply:
I think no. The idea is that CONTRIBUTING.md is enough for most if not all contributors, and this is a fall-back for cases when we need to explain why certain piece of code does not pass a code review.